rafx-framework/src/resources/descriptor_sets/descriptor_set_pool_chunk.rs

```diff
@@ -17,6 +21,453 @@ use rafx_api::{
+    descriptor_set_handle: RafxDescriptorSetHandle,
+    descriptor_set_index: usize,
+    descriptor_set_layout: ResourceArc<DescriptorSetLayoutResource>,
+    descriptor_set_array: &'a mut RafxDescriptorSetArray,
+    buffers: &'a mut DescriptorLayoutBufferSet,
+}
+
+impl DescriptorSetWriterContext<'_> {
+    pub(crate) fn handle(&self) -> RafxDescriptorSetHandle {
+        self.descriptor_set_handle.clone()
+    }
+
+    fn write(
+        &mut self,
+        element_key: &DescriptorSetElementKey,
+        element: &DescriptorSetElementWrite,
+    ) -> RafxResult<()> {
+        let descriptor_set_index = self.descriptor_set_index;
+
+        log::trace!(
+            "Process descriptor set write for {:?} {:?}. layout {:?}",
+            descriptor_set_index,
+            element_key,
+            self.descriptor_set_layout,
+        );
+
+        try_queue_descriptor_set_image_update(
+            self.descriptor_set_array,
+            descriptor_set_index as u32,
+            element_key,
+            element,
+        )?;
+
+        try_queue_descriptor_set_buffer_update(
+            self.buffers,
+            &self.descriptor_set_layout.get_raw(),
+            self.descriptor_set_array,
+            descriptor_set_index as u32,
+            element_key,
+            element,
+        )?;
+
+        Ok(())
+    }
+
+    fn write_buffer_data<T: Copy>(
+        &mut self,
+        element_key: &DescriptorSetElementKey,
+        data: &[T],
+    ) -> RafxResult<()> {
+        let descriptor_set_index = self.descriptor_set_index as u32;
+
+        log::trace!(
+            "Process descriptor set write for {:?} {:?}. layout {:?}",
+            descriptor_set_index,
+            element_key,
+            self.descriptor_set_layout,
+        );
+
+        let buffer = self
+            .buffers
+            .buffer_sets
+            .get_mut(&DescriptorSetBindingKey {
+                dst_binding: element_key.dst_binding,
+            })
+            .unwrap();
+
+        if data.len() as u32 > buffer.buffer_info.per_descriptor_size {
+            panic!(
+                "Wrote {} bytes to a descriptor set buffer that holds {} bytes layout: {:?}",
+                data.len(),
+                buffer.buffer_info.per_descriptor_size,
+                self.descriptor_set_layout
+            );
+        }
+
+        if data.len() as u32 != buffer.buffer_info.per_descriptor_size {
+            log::warn!(
+                "Wrote {} bytes to a descriptor set buffer that holds {} bytes layout: {:?}",
+                data.len(),
+                buffer.buffer_info.per_descriptor_size,
+                self.descriptor_set_layout
+            );
+        }
+
+        let offset = buffer.buffer_info.per_descriptor_stride * descriptor_set_index;
+
+        log::trace!(
+            "Writing {} bytes to internal buffer to set {} at offset {}",
+            data.len(),
+            descriptor_set_index,
+            offset
+        );
+        buffer
```

Comment:
TODO(dvd): Can this be deduplicated wrt to `try_queue_descriptor_set_buffer_update`?